compiler/includer.cc

```diff
@@ -0,0 +1,117 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "impeller/compiler/includer.h"
+
+#include "flutter/fml/paths.h"
+
+namespace impeller {
+namespace compiler {
+
+Includer::Includer(std::shared_ptr<fml::UniqueFD> working_directory,
+                   std::vector<IncludeDir> include_dirs,
+                   std::function<void(std::string)> on_file_included)
+    : working_directory_(std::move(working_directory)),
+      include_dirs_(std::move(include_dirs)),
+      on_file_included_(std::move(on_file_included)) {}
+
+// |shaderc::CompileOptions::IncluderInterface|
+Includer::~Includer() = default;
+
+std::unique_ptr<fml::FileMapping> Includer::TryOpenMapping(
+    const IncludeDir& dir,
+    const char* requested_source) {
+  if (!dir.dir || !dir.dir->is_valid()) {
+    return nullptr;
+  }
+
+  if (requested_source == nullptr) {
+    return nullptr;
+  }
+
+  std::string source(requested_source);
+  if (source.empty()) {
+    return nullptr;
+  }
+
+  auto mapping = fml::FileMapping::CreateReadOnly(*dir.dir, requested_source);
+  if (!mapping || !mapping->IsValid()) {
+    return nullptr;
+  }
+
+  on_file_included_(fml::paths::JoinPaths({dir.name, requested_source}));
+
+  return mapping;
+}
+
+std::unique_ptr<fml::FileMapping> Includer::FindFirstMapping(
+    const char* requested_source) {
+  // Always try the working directory first no matter what the include
+  // directories are.
+  {
+    IncludeDir dir;
+    dir.name = ".";
+    dir.dir = working_directory_;
+    if (auto mapping = TryOpenMapping(dir, requested_source)) {
+      return mapping;
+    }
+  }
+
+  for (const auto& include_dir : include_dirs_) {
+    if (auto mapping = TryOpenMapping(include_dir, requested_source)) {
+      return mapping;
+    }
+  }
+  return nullptr;
+}
+
+shaderc_include_result* Includer::GetInclude(const char* requested_source,
+                                             shaderc_include_type type,
+                                             const char* requesting_source,
+                                             size_t include_depth) {
```

Comment:
Is this used? Is the idea to use it to catch cycles?

Reply:
Do you mean the includer or a specific argument?

Reply:
The `include_depth` argument.

What happens if someone messes up such that file A includes file B, and file B includes file A?